node/libs/concurrency/src/sync/prunable_queue.rs

```diff
@@ -0,0 +1,148 @@
+//! Prunable queue mimics a producer-consumer channel, in enabling
+//! waiting on the receiver side for new items, while providing pruning functionality.
+//! The pruning takes place whenever a new item is sent, based on a specified predicate.
+//!
+//! The queue facilitates the asynchronous return of processing result for each added item.
+//!
+//! The separation of `Sender` and `Receiver` is employed primarily because `Receiver` requires
+//! a mutable reference to the signaling channel, unlike `Sender`, hence making it undesirable to
+//! be used in conjunction.
+
+use crate::{
+    ctx::{self, Ctx},
+    oneshot,
+    sync::{self, watch},
+};
+use std::{collections::VecDeque, sync::Arc};
+use tokio::sync::Mutex;
+
+pub fn new<T, U>(
```

Comment:
nit: the convention is to name it as "channel"

Reply:
Fixed.